### ISSUE TITLE
Render image attachments if more than 4 images are attached

### ISSUE THREAD

**Issue by ktagbnaw** [tootsuite/mastodon#14336]
### Pitch

If more than 4 images are attached to a post render the first 4 images but also keep the remote link.

### Motivation

Mastodon limits post to maximum 4 image attachments. Mastodon also doesn't render any images in incoming posts with the number of images > 4. It will show post text but no images at all, just a remote link to initial (i.e. Friendica) post at the bottom.
For non-Mastodon users that means either making only posts with 4 images each or accepting that Mastodon followers will most likely ignore their wonderful multi-image posts (people are lazy to follow remote links). For Mastodon users that means no chance to follow people who contribute a lot of beautiful photos in each post. Nothing in ActivityPub prevents more than 4 image attachments.

Suggestion: keep the remote link to post AND show the first 4 images in incoming posts with >4 images.

Issue raised by https://mastodon.xyz/@lightone/104519548954134472

**Comment by ThibG** [comment]
Hm, what happens is actually a bit different. Mastodon doesn't reject messages with more than 4 attachments, nor does it drop all attachments at all. But it keeps only the first 4, which is a deliberate decision (although I do not agree with it, I'd prefer a much higher hard limit, and some UI tweaks to allow more than 4 attachments)

**Comment by Sasha-Sorokin** [comment]
As temporary solution, maybe we could add link below the media gallery that says like “2 more images in the source”, sort of we did with unfetchable followers or old posts.

<details>
<summary>Pseudocode</summary>

```jsx
{attachmentsCount > ATTACHMENTS_LIMIT
  ? <a href={statusURL} className='iconic_link' ref='noreferrer noopener'>
      <Icon id='picture' />
      <FormattedMessage
        id='status.more_attachments'
        defaultMessage='{count, plural, one {# more image} other {# more images}} in the source'
        values={{ count: ATTACHMENTS_LIMIT - attachmentsCount }}
      />
    </a>
  : null}
```
</details>

**Comment by ThibG** [comment]
Yes, something like that (even though my proposal would be more along the lines of having a “…”/“and N more images” over the preview widget, and all of them viewable in the modal). But your change isn't enough as the original number of attachments isn't saved (they are discarded server-side).

**Comment by Sasha-Sorokin** [comment]
@ThibG wdym “over the preview widget”?

> they are discarded server-side

Ah, snap. Server said no fun allowed 😩

**Comment by Sasha-Sorokin** [comment]
[@lightone@mastodon.xyz](https://mastodon.xyz/@lightone) seem to have lost access to her GitHub account, but wanted to comment on this issue too:

> @ThibG [\[that\]](https://github.com/tootsuite/mastodon/issues/14336#issuecomment-659398635) isn't quite right. When I send a post with 5 or more images from Friendica, I see NO IMAGES at all in Mastodon. The post kind of shrinks, shows me only the text and a link at the end to whole remote post. That's the point of the issue!

Updated later:

> I have an update to the issue. I've just checked, @ThibG is right. *Today* if Mastodon gets a post with >4 images from Friendica, it shows first 4 images. 2 weeks ago, when I tested from same Friendica node and when the issue was opened, the result was different! Mastodon received such posts and rendered no images at all. BUT it showed text and added a link to original remote post at the end of post.

> [Now] Mastodon DOES NOT add any remote link, just shows first 4 images and discards the rest. I'm confused. 😳 Perhaps some related code was changed either in Friendica or in Mastodon during these 2 weeks?
> 
> So partly my issue was resolved (somehow 🤷 ). First 4 images ARE shown in many-image posts. So indeed the only improvement now would be to add some note at the end of such posts in Mastodon: "more images in /original link to post/"

Note: **[@lightone](https://mastodon.xyz/@lightone)** isn't the poster of issue :)

**Comment by trwnh** [comment]
It would be nice to at least see *some* indication that media exists on the remote, even if it is rejected locally. Right now, the lack of indication that more attachments exist amounts to a lack of context.

This doesn't happen with just images, either -- I saw a post that had two audio attachments show up in my feed with only one attachment, which changed the meaning of the post. The same thing could happen if e.g. someone posted a 5-part comic where the 5th panel/upload contained information that majorly changes the intent of the first 4 alone.

**Comment by tribela** [comment]
> Yes, something like that (even though my proposal would be more along the lines of having a “…”/“and N more images” over the preview widget, and all of them viewable in the modal). But your change isn't enough as the original number of attachments isn't saved (they are discarded server-side).

I am longing for this feature. Cutting up to 4 images without leaving any info is very bad IMO.

**Comment by rokejulianlockhart** [comment]
> I'd prefer a much higher hard limit

I don't think we should limit it.

**Comment by ThisIsMissEm** [comment]
This seems to be a dupe of https://github.com/mastodon/mastodon/issues/10258 or vice-versa.

**Comment by trwnh** [comment]
#10258 is about an indicator of more images, while this one is about showing the images

**Comment by sinoru** [comment]
Is it possible to keep all images, but limit 4 on local upload and showing preview?

I tried remove limitation from ActivityPub parsing parts (https://github.com/sinoru/mastodon/commit/206943b1e69dc2226a72466e31ef365bbbef1440), Mastodon can get all images, web only shows 4 previews, but when go in details, It can swipe to more.

<img width="1343" alt="image" src="https://github.com/mastodon/mastodon/assets/1138433/eda1d61a-ee08-4181-8a38-018a4295cce9">
<img width="1343" alt="image" src="https://github.com/mastodon/mastodon/assets/1138433/676c001c-4abe-42f3-9aa8-9306b966af4f">

**Comment by ShadowJonathan** [comment]
I think an absolute maximum 32 images (to prevent DoS) would be more than sufficient, tbh.

**Comment by Shanesan** [comment]
This is an incredibly Twitter-style restriction that needs to be alleviated per-server. Essentially Mastodon is playing with hard-set rules in a fediverse that has incredibly loose rules. 

It’s time for the option to change.

**Comment by Shanesan** [comment]
@sinoru has this been working out for you? If so, I hope you can submit a pull request to the main branch.

**Comment by FediVideos** [comment]
Had a lot of new people asking via FediTips about why Mastodon isn't showing their complete Pixelfed posts.

Is there any chance of Mastodon reconsidering its decision to only display four images? The four image restriction is causing confusion, making people wonder if Mastodon even federates properly as the images are often missing.

**Comment by nclm** [comment]
This is becoming more and more crucial with Pixelfed’s renewed popularity.

It happens to me daily now: look at 4 pictures, don’t understand the point of the poster, prepare to scroll past, decide instead to have an extra look, check the poster’s server, realise it’s a Pixelfed server, find my way to the original post, and finally get the post. 

It’s a very bad UX, and it’s only because I kinda know a little about Mastodon and Pixelfed, and only on posts in which the pictures carry some meaning together (comics, etc), that I can figure my way. Most newer users would see 4 pictures, never understand what’s wrong, and scroll past.

Showing all the pictures or at least adding a mention of more pictures at the link seems essential at that point.

**Comment by Clemovski** [comment]
Hello.
Can someone elaborate on where this is going right now please ?
This pull request https://github.com/mastodon/mastodon/pull/29562 is one year old. Is it abandoned ? Is the current issue still being worked on ?
Thank you.

**Comment by wolfspyre** [comment]
I would like to amplify the challenge this somewhat arbitrary limitation presents.

what can be done to get some traction here?

**Comment by Shanesan** [comment]
> Hm, what happens is actually a bit different. Mastodon doesn't reject messages with more than 4 attachments, nor does it drop all attachments at all. But it keeps only the first 4, which is a deliberate decision (although I do not agree with it, I'd prefer a much higher hard limit, and some UI tweaks to allow more than 4 attachments)

@ClearlyClaire In your quote above where you state it's a deliberate decision, how/where do we bring this up to review this deliberate (and I don't agree with it either) decision?

**Comment by wolfspyre** [comment]
The real kicker, as mentioned [here](https://github.com/mastodon/mastodon/issues/18127#issuecomment-2946421341) is that Mastodon does this **SILENTLY** 
the viewing user has ABSOLUTELY NO SIGNAL that they are not viewing all content from the original post..  

How can this possibly be considered a good user experience? 

I concede that given the history here, changing the UI/related views/other facets of masto may be non-trivial.. 

Nevertheless, the current practice is essentially an arbitrary decision to just drop source content and sweep it under the rug  **SILENTLY**.

I think it's safe to say this is NOT the right way forward. 

If Mastodon still has a bee in its' bonnet about message size, styling, or.... 
_I do not know what the original reasoning for this behavior was when this decision was made, so I can't really speak to its current relevance today..._  

... and **even in the light of the bad ux that this practice represents**... still believes the right thing to do for the masto user is to only present the user with up to 4 distinct media assets... 

I'd assert that the bare minimum forward here is to add some sort of "... **N** additional (images/vides/documents/pictograms/songs/horrible_memes/etc) not shown ... click here to view the post on _insert-platform-emitting-source-content-here_ ) 
blurb to the post in question.... 

at the very least, INFORM the user that you've made the decision that these are the only pieces of the original content they need to see as proposed in https://github.com/mastodon/mastodon/issues/10258

PLEASE?!
